zeroeq/receiver.cpp

```diff
@@ -96,51 +96,71 @@ class Receiver
         }
     }
 
-    bool _receive( const uint32_t timeout )
+    bool _receive( uint32_t timeout )
     {
-        std::vector< Socket > sockets;
-        std::deque< size_t > intervals;
-        for( ::zeroeq::Receiver* receiver : _shared )
+        // ZMQ notifications on its sockets is edge-triggered, hence we have
+        // to receive all pending POLLIN events to not 'loose' notifications
+        // from the socket descriptors (c.f. HTTP server).
+        // For reference:
+        // https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification
+        bool haveData = false;
+        do
         {
-            const size_t before = sockets.size();
-            receiver->addSockets( sockets );
-            intervals.push_back( sockets.size() - before );
-        }
+            std::vector< Socket > sockets;
+            std::deque< size_t > intervals;
+            for( ::zeroeq::Receiver* receiver : _shared )
+            {
+                const size_t before = sockets.size();
+                receiver->addSockets( sockets );
+                intervals.push_back( sockets.size() - before );
+            }
 
-        switch( zmq_poll( sockets.data(), int( sockets.size( )),
-                          timeout == TIMEOUT_INDEFINITE ? -1 : timeout ))
-        {
-        case -1: // error
-            ZEROEQTHROW( std::runtime_error( std::string( "Poll error: " ) +
-                                             zmq_strerror( zmq_errno( ))));
-        case 0: // timeout; no events signaled during poll
-            return false;
+            switch( zmq_poll( sockets.data(), int( sockets.size( )),
+                              timeout == TIMEOUT_INDEFINITE ? -1 : timeout ))
+            {
+            case -1: // error
+                ZEROEQTHROW( std::runtime_error( std::string( "Poll error: " ) +
+                                                 zmq_strerror( zmq_errno( ))));
+            case 0: // timeout; no events signaled during poll
+                return haveData;
```

Comment:
Not entirely sure too. What I was trying to determine is if 'case 0' can ever happen after the first call. It seems to me that since we only do another zmq_poll if(socket.revents & ZMQ_POLLIN), then the poll can't possibly return 0 in that case and that if it does, then that is a unexpected beahviour, not a regular return true.